### DISCUSSION TITLE
Reclaim your REV Expansion Hubs from your Outreach Bots

### DISCUSSION THREAD

**Discussion by AlecHub** [FIRST-Tech-Challenge/FtcRobotController#588] [Show and tell] [answered]
A [REV Expansion Hub](https://www.revrobotics.com/rev-31-1153/) (“Hub”) can be reclaimed from your outreach bot(s) by using inexpensive COTS alternatives.  The [REV SPARKmini](https://www.revrobotics.com/rev-31-1230/) and [REV Servo Power Module](https://www.revrobotics.com/rev-11-1144/) can also be reclaimed. 

| Reclaimed Hubs and other devices can be sold or donated to other teams, or re-used on the following season's competition bot. |
| :--- |

A Hub can be reclaimed for as little as $25 since typically many of the ports on a Hub are not needed.  In the case where all 4 motor ports and/or all 6 servo ports are needed, a Hub can be reclaimed for as little as $55.

COTS devices can also be used in the development stages of the competition bot to minimize the chances of damaging a Hub during development.

Additionally, COTS devices allow teams to implement and demo features not possible to implement with the Hub, which can impress judges during judging.

Perhaps the greatest value of using COTS devices is that students gain valuable marketable skills by way of hands-on experience with devices used in the industry.
## COTS Alternatives
The critical components of the Hub are the four (4) DC Motor ports and the six (6) Servo ports.  The following are COTS alternatives:
- PCA9685 Breakout Board (as little as $6 per board)
- PCA9685 compatible motor driver(s) (as little as $12 per channel)
> [!Note] 
> Unless otherwise noted, all COTS devices are compatible with 3.3V and 5V power or logic level.
### PCA9685 Breakout Board 
A [NXP PCA9685](https://www.nxp.com/docs/en/data-sheet/PCA9685.pdf) Breakout Board (“PCA9685”) provides 16 PWM ports.  These ports can be used to control:
- up to 16 Servos, 
- up to 16 DC Motors, or
- a combination of up to 16 Servos and DC Motors

Due to the wide variety of uses of the PCA9685, numerous manufacturers have produced breakout boards:

![PCA9685](https://user-images.githubusercontent.com/15099062/213499393-4879f36b-032c-47d5-a0b4-ae08518a953f.jpg)

The breakout boards are available from numerous vendors.
# Motor Drivers
A DC Motor Driver (“Motor Driver”) has one or more Motor Channels, and one or more Input Interface options per channel.  Common Input Interface options are:
- USB 
- UART 
- Analog 
- I2C
- Servo 
- PWM 

Many Motor Drivers provide a “Servo Interface”.  A Servo Interface is a PWM protocol used by devices commonly referred to as “RC Servo”, “Hobby Servo”, and/or “3-Wire Servo”.

Other Motor Drivers provide a “PWM Interface”.  A PWM Interface has a different PWM protocol and wiring than the Servo Interface.

The PCA9685 supports the Servo Interface and the PWM Interface.

## Motor Channel
A Motor Driver powers a DC Motor via the 12V robot main battery.  The Motor Driver uses the PWM technique to regulate the voltage to a Motor Channel.  The effective voltage to a motor connected to a Motor Channel is determined by the duty cycle of the PWM pulses.  The no-load RPM of the motor is determined by this voltage.

## Input Interfaces
There two classes of PWM interfaces on input to a Motor Driver:
- Servo Interface
- PWM Interface

### Servo Interface
The Servo Interface has one analog control signal.  The signal is a PWM protocol.  Nominally, the signal has:
- PWM frame period of 20 ms
- Pulse width of 1.0 to 2.0 ms
- Pulse amplitude of 3.0 to 5.0 V

Typically, the PWM frame frequency of the Servo Interface is 50 Hz (20 ms duration).  Although 50 Hz is the nominal frequency, Servos and Motor Drivers support a range of input frame frequencies.  Therefore, the duty cycle of the Servo Interface, has no semantics.  The semantics are in the pulse widths.

The pulse width of the Servo Interface determines the duty cycle and polarity of the PWM of the Motor Channel (i.e. RPM and direction of rotation).  The frame frequency of a Motor Channel is vendor specific and is fixed to some frequency, typically between 10,000 and 20,000 Hz.

The following motor controllers provide a Servo Interface:
- [REV SPARKmini](https://www.revrobotics.com/rev-31-1230/)
- [VEX Motor Controller 29](https://www.vexrobotics.com/276-2193.html)
- [goBILDA 1x15A Motor Controller](https://www.gobilda.com/1x15a-motor-controller-30a-peak/) 

### PWM Interfaces
The semantics of the PWM Interfaces is simple: The duty cycle of the PWM Interface determines the duty cycle of the Motor Channel.  The PWM of the Motor Channel has the exact same frame frequency, duty cycle, and pulse timing as the input PWM Interface.  

The key information carried by the interfaces are:
- **Servo Interface:**  Speed and Direction of Rotation of the Motor
- **PWM Interfaces:**  Duty Cycle and Timing of Motor Channel pulses

Thus, the PWM Interfaces allow the control program to have direct fine-grained control of the PWM of a Motor Channel.

The drawback of the PWM Interfaces is that direction of rotation information must get sent out-of-band.  Therefore, Motor Drivers having a PWM Interface have two input signals.

#### PWM+DIR Interface
The PWM+DIR Interface has a PWM input and a DIR input.  Direction of rotation is controlled by the DIR input, which is just a Digital Input channel.

PWM and DIR are connected to two channels of the PCA9685.  Any PCA9685 channel can be configured as a Digital Output channel.

#### PWM+PWM Interface
The PWM+PWM Interface has two PWM inputs, say “PWM1” and “PWM2”.  These are intended to drive a phase of a Stepper Motor but can be used to drive a DC Motor as well.

Sending a PWM signal on PWM1 only will turn a DC motor in one direction, and sending the signal on PWM2 only will turn the motor in the other direction.

PWM1 and PWM2 are connected to two channels of the PCA9685.

## PCA9685 16-channel I2C LED Controller
Although the PCA9685 is a LED Controller, it can be configured to support a Servo Interface or a PWM Interface.

There is one key aspect of the PCA9685 to keep in mind.  All 16 channels operate on the same chip-wide frame frequency.  Typical frame frequencies and their applications are:
- **50 Hz:**  RC/Hobby Servo, or Motor Driver with a Servo Interface
- **100 Hz:**  LEDs
- **1000 Hz:**  Motor Driver with a PWM Interface

A separate PCA9685 board is needed for each type of application:

![DSCN9084-6](https://user-images.githubusercontent.com/15099062/215797743-65e6013a-4eb6-4438-9470-c1513517ec97.jpg)

The following I2C bus modes are supported by the PCA9685:
- **Standard-mode (Sm)**: 100 kbit/s
- **Fast-mode (Fm)**: 400 kbit/s 
- **Fast-mode Plus (Fm+)**: 1 Mbits/s

> [!Note]
> **SDA** and **SCL** on the PCA9685 have **10 kΩ** pullups to VCC.  The Hub has **2.49 kΩ** pullups to VCC. Thus, connecting 3 PCA9685 boards to a I2C port on a Hub yields **1.43 kΩ** pullups to VCC for the bus.
# **_FIRST_<sup>®</sup> ENERGIZE<sup>℠</sup>**
A key feature of the PCA9685 is Pulse Delay.  Each channel can be configured with a different delay to the start of the pulse from the start of the frame.

If you have two drive motors, you would delay one of the channels by 50% of the frame duration.  When the duty cycle is less than 50% on each channel, the robot main battery is not having to send power to both motors at the same time.  Otherwise when the duty cycle is greater than 50% but less than 100%, the duration that the battery has to power both motors at the same time is minimized.

Likewise if you have four drive motors, you would delay three of the four drive motor PWM channels by 25%, 50%, and 75% of the frame duration.

Or you could implement more elaborate algorithms where the delay of each channel is a function of the duty cycle of all channels at any given time.  For example, if one channel is at 30% and another channel is at 70%, you could dynamically delay the second channel by 30% of the frame duration so that the main battery does not have to power both motors at the same time.

The Pulse Delay feature is only effective for PWM Interfaces since the timing of Motor Channel pulses cannot be controlled via the Servo Interface.  Therefore, we prioritized testing motor drivers which have a PWM+DIR Interface or a PWM+PWM Interface.

# Tested Motor Drivers
Available from a variety of vendors:

![DSCN9097-A](https://user-images.githubusercontent.com/15099062/213896527-fe999441-2ed9-42e1-9a56-9b27541688b5.jpg)
## Cytron
The following Cytron motor drivers have a PWM+DIR Interface: 
| Model | Channels | Continuous Current (A) | Peak Current (A) | Price |
| ------- | :--------: | :--------: | :--------: | :--------: |
| MDD10A | 2 | 10 | 30 | $24 |
| MD10C | 1 | 13 | 30 | $14 |
| MD13S | 1 | 13 | 30 | $14 |

The Cytron motor drivers will Brake the motor when the duty cycle is 0%. 
## DBH-12V
The unbranded DBH-12V Dual-Channel H-Bridge Stepper Driver has a PWM+PWM Interface typical of stepper motor drivers:
| Model | Channels | Continuous Current (A) | Peak Current (A) | Price |
| ------- | :--------: | :--------: | :--------: | :--------: |
| DBH-12V | 2 | 20 | 30 | $25 |

### Connections
![DBH-12V-2](https://user-images.githubusercontent.com/15099062/214104003-6c6b33cb-a44e-4541-898f-9a47a5d09cf5.jpg)

> [!Warning] 
> **V+** is connected to **PWR+** (i.e., battery voltage).  To prevent mishaps, you could cut the two V+ pins off the board

**IN1 & IN2** pins are for the two PWM input signals.  Only send a PWM signal on IN1 or IN2; not to both at the same time.  Duty cycle of the PWM signal must be less than 98% for the DBH-12V (because a waveform needs to be present?). 

**EN** is a digital input pin for enabling and disabling the H-Bridge.  The H-Bridge is enabled when nothing is connected to EN.  When the H-Bridge is enabled, the motor will Brake on 0% duty cycle.  Connecting EN to a PCA9685 channel allows the control program to enable/disable the H-Bridge.  When the H-Bridge is disabled, the motor will Coast.

**CT** is an analog output pin (0 to 5V).  The voltage is representative of the current draw of the motor.  

**A** & **B** terminals are the two output Motor Channels.

> [!Note] 
> **V+** pins _could_ be used to power low power motors such as the VEX 393 Motor:
>
> ![DSCN9193-2](https://user-images.githubusercontent.com/15099062/215160529-85653607-7bf7-4679-b4cd-55bc643419c7.jpg)
> _The **VCC** bus on the PCA9685 powers the NXP PCA9685 chip, which generates the PWM signals.  The **V+** bus supplies power to devices connected to the 3-pin PWM ports.  The two busses are isolated from each other and can be different voltages._

> [!Warning] 
> Typically you shouldn't pass V+ from board to board via the V+ pin of the PCA9685 as shown above.  Instead you should power each board independently as needed via the V+ terminal.
# Judging
The emphasis of **_FIRST_<sup>®</sup> Tech Challenge** is on the judging aspect of the competition.  Reclaiming a Hub from an outreach bot allows teams to:
- Help a team in need of a Hub while significantly reducing the cost of the outreach bot
- Learn new motor driver interfaces and protocols
- Learn how to write a I2C driver and motor control software
- Learn important roles that PWM plays in electronics and motion control systems
- Implement the **_FIRST_<sup>®</sup> ENERGIZE** and **INNOVATE** themes in meaningful and substantive ways

Additionally, students can earn the fundamentals of LED control and demonstrate the same on a competition bot.

All are great ways of impressing the judges.  Moreover, students obtain valuable marketable skills by way of hands-on experience with COTS devices and technologies used in the industry.
> [!Note] 
> Per GM1 it is perfectly legal to use a PCA9685 for competition, however you are only allowed to use it to control LEDs:
>
> ![LEDS-4](https://user-images.githubusercontent.com/15099062/215831903-0a4e0d72-6501-4a33-8f5b-7af35394ae0d.JPG)
> _Four Common Cathode RGB LEDs connected to 12 channels of a PCA9685.  The PCA9685 limits the current draw of each channel to 20 mA._ Frame frequency is 100 Hz.
# Coming Soon
Thanks to the **[OctoQuad](https://www.tindie.com/products/digitalchickenlabs/octoquad-8ch-quadrature-pulse-width-decoder/)** from @DigitalChickenLabs, you can reclaim **_both_** REV Control Hub **_and_** REV Expansion Hub for as little as $150. 

![OctoQuad](https://user-images.githubusercontent.com/15099062/215829731-6dfa5c97-30f5-49b0-8537-84e809694a5c.jpg)

A fantastic project for the pre-season.   **_PLUS_** much more!

**Comment by gearsincorg** [comment, marked as answer]
A Brillant idea!  

Our outreach robot (demobot) is always the first to suffer when we need spare parts.

Glad you included the OctoQuad.  It can be used (out of the box) to add encoders to a ControlHub in a outreach bot.  
This lets you keep your Motor and Odometry encoders connected.  

There's even a tutorial for doing it here: https://github.com/DigitalChickenLabs/OctoQuadFtcDriver/tree/master/tutorials

**Comment by texasdiaz** [comment]
I presume you're using the REV Control Hub or some other platform with an I2C interface to drive this stack? Do you have sample drivers for this?

**Reply by AlecHub** [reply]
Good questions... I've tested all three PCA9685 boards depicted in my images (addresses 0x42, 0x43, 0x44) connected to:
- One of the I2C busses of the Control Hub
- A **USB to I2C Adapter** connected to:
  - A USB port of a Control Hub
  - The USB port of a Robot Controller Phone
  - A USB Hub connected to:
    - A USB port of a Control Hub
    - The USB port of a Robot Controller Phone
  - A USB Hub with other connected **USB to I2C Adapters** connected to:
    - A USB port of a Control Hub
    - The USB port of a Robot Controller Phone

Yes, a Robot Controller Phone that has multiple I2C busses (and SPI busses).

Any **USB to I2C Adapter** should work. I've tested several **CH341** and **FT232H** based adapters. I have no doubt that the **CP2112** based adapters will also work flawlessly.

Thank you @3cky for your awesome **[usb-i2c-android](https://github.com/3cky/usb-i2c-android)** library!!! It is a pleasure to work with your API, especially the `UsbI2cManager.getAdapters()` enumeration function.

As for the PCA9685 driver, there are hundreds of interface classes on Github, some even in java.  I feel it would be a beneficial exercise for students to port one of those classes to a `I2cDeviceSynchDevice<I2cDeviceSynch>` driver.

Coming Soon: **USB to SPI**  via the same multi-protocol **USB to I2C Adapters** mentioned above.
